python/pyxet/pyxet/file_operations.py

```diff
@@ -0,0 +1,329 @@
+    if _isdir(dest_fs, dest_path): 
+        # Split up the final component from source path and add it
+        # to the destination.
+        # We get final component from 'source' instead of the 
+        # normalized 'src_path' because if 'source' ends with 
+        # '/' or '\', the desired behavior is to copy what's 
+        # under that directory, and to skip 'source' itself.
+        final_source_component = _path_split(src_fs, src_path)
+        file_dest_path = _path_join(dest_fs, dest_path, final_source_component)
+        cp_unit = CopyUnit(src_path=src_path, dest_path=file_dest_path, dest_dir = None, size = size_hint)
+    else:
+        dest_dir = _path_split(dest_fs, dest_path)
+        if _isdir(dest_fs, dest_dir):
+          dest_dir = None # No need to create this
+
+        cp_unit = CopyUnit(src_path=src_path, dest_path=dest_path, dest_dir = dest_dir, size = size_hint)
+                           
+    _single_file_copy_impl(cp_unit, src_fs, dest_fs)
+
+
+def _build_cp_action_list_impl(src_fs, src_path, dest_fs, dest_path, recursive):
+    
+    dest_is_xet = dest_fs.protocol == "xet"
+    # If both the source and the end is a xet, then this can be done 
+    dirs_directly_copyable = dest_is_xet and src_fs.protocol == "xet"
+
+    dest_has_trailing_slash = dest_path.endswith("/") 
+    dest_path = dest_path.rstrip("/")
+
+    # Handling wildcard cases
+    if '*' in src_path:
+
+        # validate
+        # we only accept globs of the for blah/blah/blah/[glob]
+        # i.e. the glob is only in the last component
+        # src_root_dir should be blah/blah/blah here
+        src_root_dir, end_component = _path_split(src_fs, src_path)
+
+        if '*' in src_root_dir:
+            raise ValueError(
+                f"Invalid glob {src_path}. Wildcards can only appear in the last position")
+
+        src_path = src_root_dir 
+        glob_pattern = end_component if end_component != "*" else None
+
+        # Ensure we have the right behavior here below.
+        src_path_trailing_slash = True
+        dest_has_trailing_slash = True
+
+        dest_dir=dest_path
+        dest_path=None # In this case, the path information will be taken from the source file name.
+        src_isdir=True
+
+    else:
+        glob_pattern = None
+
+        src_path_trailing_slash = src_path.endswith("/")
+        
+        if src_path_trailing_slash:
+            src_path = src_path.rstrip("/")
+            src_isdir = True
+            if not _isdir(src_fs, src_path): 
+                raise ValueError(f"Source path {src_path} not an existing directory.")
+        else:
+            src_isdir = _isdir(src_fs, src_path)
+
+        # Handling directories
+        if src_isdir:
+            if not recursive:
+                raise ValueError(
+                    "Specify recursive flag '-r' to copy directories.")
+
+        # Figure out what the destination is like, etc. 
+        try:
+            info = dest_fs.info(dest_path)
```

Comment:
Since info is only used to check whether the path is directory-like, we can just reuse the `_isdir` function